File: day_1/java/src/me/h2ofiremaster/Part2.java
package me.h2ofiremaster;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Locale;
import java.util.Map;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Part2 {
    private static final File INPUT_FILE = new File("input.txt");
    private static final Map<String, String> NUMBER_MAP = Map.of(
            "one", "1",
            "two", "2",
            "three", "3",
            "four", "4",
            "five", "5",
            "six" ,"6",
            "seven", "7",
            "eight", "8",
            "nine", "9"
    );
    private static final Pattern NUMBER_REGEX = Pattern.compile("(?=("
            + Stream.concat(NUMBER_MAP.keySet().stream(), NUMBER_MAP.values().stream())
                    .collect(Collectors.joining("|"))
            + "))");
    
    
    public static void main(String[] args) {
        Scanner fileScanner;
        try {
            fileScanner = new Scanner(INPUT_FILE);
        } catch(FileNotFoundException e) {
            System.err.println("File '" + INPUT_FILE + "' was not found.");
            return;
        }
        
        int output = 0;
        while(fileScanner.hasNextLine()) {
            String line = fileScanner.nextLine();
            line = convertToNumbers(line);
            
            if(line.length() < 1) continue;
            String number = String.valueOf(line.charAt(0)) + line.charAt(line.length() - 1);
            
            try {
                output += Integer.parseInt(number);
            } catch(NumberFormatException e) {
                System.err.println("'" + number + "' is not a valid number.");
            }
        }
        
        System.out.println(output);
    }
    
    public static String convertToNumbers(String line) {
        String input = line.toLowerCase();
        Matcher matcher = NUMBER_REGEX.matcher(input);
        return matcher.results()
                .map(r -> r.group(1))
                .map(r -> NUMBER_MAP.getOrDefault(r, r))
                .collect(Collectors.joining(""));
    }
}

File: day_1/c++/Part2.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <map>


using std::string;
using std::regex;

const std::map<std::string, char> NUMBER_MAP = {
    {"one", '1'},
    {"two", '2'},
    {"three", '3'},
    {"four", '4'},
    {"five", '5'},
    {"six", '6'},
    {"seven", '7'},
    {"eight", '8'},
    {"nine", '9'}
};

const string INPUT_FILE = "input.txt";

inline regex generateNumbersRegex() {
    string numbersRegex = "(?=(";
    
    for(const auto &element : NUMBER_MAP)
        numbersRegex += element.first + '|' + element.second + '|';

    numbersRegex.pop_back();
    numbersRegex += "))";

    return regex(numbersRegex);
}
const regex NUMBERS_REGEX = generateNumbersRegex();

string convertToNumbers(string input) {
    std::transform(input.begin(), input.end(), input.begin(), ::tolower);
    string output;
    std::sregex_iterator numbers(input.begin(), input.end(), NUMBERS_REGEX);
    std::sregex_iterator end;

    for(; numbers != end; numbers++) {
        if((*numbers).size() < 2) continue;

        string number = (*numbers)[1].str();

        if(number.size() == 1)
            output += number;
        else
            output += NUMBER_MAP.at(number);
    }
    
    return output;
}

int main() {

    int output = 0;
    string line;
    std::ifstream inputFile(INPUT_FILE, std::ios::in);
    if(inputFile.is_open()) {
        while(std::getline(inputFile, line)) {
            
            line = convertToNumbers(line);
            
            char firstChar = line[0];
            char lastChar = line[line.length() - 1];
            
            int firstInt = firstChar - '0';
            int lastInt = lastChar - '0';
            output += (firstInt * 10) + lastInt;
        }
    } else {
        std::cout << "Unable to open file: " << INPUT_FILE << std::endl;
        return -1;
    }
    
    std::cout << output << std::endl;
    return 0;
}